// @mintlify/prebuild/dist/generate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.optionallyAddLeadingSlash = exports.slugToTitle = exports.generateDecoratedMintNavigationFromPages = void 0;
var generateDecoratedMintNavigationFromPages = function (filenamePageMetadataMap, mintConfigNav) {
    var filenames = Object.keys(filenamePageMetadataMap);
    var createNav = function (nav) {
        return {
            group: nav.group,
            version: nav === null || nav === void 0 ? void 0 : nav.version,
            pages: nav.pages.map(function (page) {
                if (typeof page === 'string') {
                    if (filenames.includes(page)) {
                        return filenamePageMetadataMap[page];
                    }
                    else {
                        console.log("\u26A0\uFE0F   \"".concat(page, "\" is defined in the mint.json navigation but the file does not exist."));
                        return {
                            title: (0, exports.slugToTitle)(page),
                            href: optionallyAddLeadingSlash(page),
                        };
                    }
                }
                return createNav(page);
            }),
        };
    };
    return mintConfigNav.map(function (nav) { return createNav(nav); });
};
exports.generateDecoratedMintNavigationFromPages = generateDecoratedMintNavigationFromPages;
var slugToTitle = function (slug) {
    var slugArr = slug.split('/');
    var defaultTitle = slugArr[slugArr.length - 1].split('-').join(' '); //replace all dashes
    defaultTitle = defaultTitle.split('_').join(' '); //replace all underscores
    defaultTitle = defaultTitle.charAt(0).toUpperCase() + defaultTitle.slice(1); //capitalize first letter
    return defaultTitle;
};
exports.slugToTitle = slugToTitle;
function optionallyAddLeadingSlash(path) {
    if (path.startsWith('/')) {
        return path;
    }
    return '/' + path;
}
exports.optionallyAddLeadingSlash = optionallyAddLeadingSlash;
